torch/_higher_order_ops/auto_functionalize.py

```diff
@@ -234,23 +361,82 @@ def do_auto_functionalize(
             normalized_kwargs[arg.name] = arg.default_value
 
     unwrapped_kwargs = ctx.unwrap_tensors(normalized_kwargs)  # type: ignore[arg-type]
+
     if "self" in unwrapped_kwargs or "self_" in unwrapped_kwargs:
         warnings.warn(
             "Using `self` or `self_` as an argument in the definition of custom ops may lead to ambiguous parsing. "
             "Please consider using a different name for this argument to avoid potential issues."
         )
+    # List of the name of args that get mutated (according to the schema)
+    mutable_args_names, mutable_args_types = get_mutable_args(op)
+
+    # A list of all bases of mutable args without duplication
+    all_basis = []
+    all_basis_addresses: list[int] = []
+
+    # Map arg_name to the index of its base in all_basis.
+    arg_to_base_index: Dict[str, Any] = {}
+
+    def update_dict(tensor, arg_name, index=None):
+        base = tensor if tensor._base is None else tensor._base
+
+        def set_result(base_index):
+            if index is None:
+                arg_to_base_index[arg_name] = base_index
+            else:
+                arg_to_base_index[arg_name][index] = base_index
+
+        if not all_basis_addresses.__contains__(base._cdata):
+            all_basis_addresses.append(base._cdata)
+            all_basis.append(base)
+            set_result(len(all_basis) - 1)
+        else:
+            set_result(all_basis_addresses.index(base._cdata))
+
+    for arg_name in mutable_args_names:
+        arg = normalized_kwargs[arg_name]
+        if arg is None:
+            continue
+
+        if isinstance(arg, list):
+            arg_to_base_index[arg_name] = {}
+            for i, tensor in enumerate(arg):
+                if tensor is None:
+                    arg_to_base_index[arg_name].append(None)
+                    continue
+
+                update_dict(tensor, arg_name, i)
+
+        else:
+            update_dict(arg, arg_name)
+
+    # add view_meta for each args into unwrapped_kwargs.
+    serialize_views_meta(
+        mutable_args_names,
+        mutable_args_types,
+        normalized_kwargs,
+        unwrapped_kwargs,
+        arg_to_base_index,
+    )
```

Comment:
Is this mutating unwrapped_kwargs? Can we instead have `serialize_views_meta` return the new view_meta and then handle the mutation of unwrapped_kwargs outside of it? Side effects in helper functions are difficult to reason about.

Reply:
discussed this point offline